ext/include/opentelemetry/ext/http/client/nosend/http_client_nosend.h

```diff
@@ -0,0 +1,184 @@
+#  include "gmock/gmock.h"
+
+using namespace testing;
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace ext
+{
+namespace http
+{
+namespace client
+{
+namespace nosend
+{
+
+const opentelemetry::ext::http::client::StatusCode Http_Ok = 200;
+
+class Request : public opentelemetry::ext::http::client::Request
+{
+public:
+  Request() : method_(opentelemetry::ext::http::client::Method::Get), uri_("/") {}
+
+  void SetMethod(opentelemetry::ext::http::client::Method method) noexcept override
+  {
+    method_ = method;
+  }
+
+  void SetBody(opentelemetry::ext::http::client::Body &body) noexcept override
+  {
+    body_ = std::move(body);
+  }
+
+  void AddHeader(nostd::string_view name, nostd::string_view value) noexcept override
+  {
+    headers_.insert(std::pair<std::string, std::string>(static_cast<std::string>(name),
+                                                        static_cast<std::string>(value)));
+  }
+
+  void ReplaceHeader(nostd::string_view name, nostd::string_view value) noexcept override;
+
+  virtual void SetUri(nostd::string_view uri) noexcept override
+  {
+    uri_ = static_cast<std::string>(uri);
+  }
+
+  void SetTimeoutMs(std::chrono::milliseconds timeout_ms) noexcept override
+  {
+    timeout_ms_ = timeout_ms;
+  }
+
+public:
+  opentelemetry::ext::http::client::Method method_;
+  opentelemetry::ext::http::client::Body body_;
+  opentelemetry::ext::http::client::Headers headers_;
+  std::string uri_;
+  std::chrono::milliseconds timeout_ms_{5000};  // ms
+};
+
+class Response : public opentelemetry::ext::http::client::Response
+{
+public:
+  Response() : status_code_(Http_Ok) {}
+
+  virtual const opentelemetry::ext::http::client::Body &GetBody() const noexcept override
+  {
+    return body_;
+  }
+
+  virtual bool ForEachHeader(
+      nostd::function_ref<bool(nostd::string_view name, nostd::string_view value)> callable)
+      const noexcept override;
+
+  virtual bool ForEachHeader(
+      const nostd::string_view &name,
+      nostd::function_ref<bool(nostd::string_view name, nostd::string_view value)> callable)
+      const noexcept override;
+
+  virtual opentelemetry::ext::http::client::StatusCode GetStatusCode() const noexcept override
+  {
+    return status_code_;
+  }
+
+public:
+  Headers headers_;
+  opentelemetry::ext::http::client::Body body_;
+  opentelemetry::ext::http::client::StatusCode status_code_;
+};
+
+class HttpClient;
+
+class Session : public opentelemetry::ext::http::client::Session
+{
+public:
+  Session(HttpClient &http_client,
+          std::string scheme      = "http",
+          const std::string &host = "",
+          uint16_t port           = 80)
+      : http_client_(http_client), is_session_active_(false)
+  {
+    host_ = scheme + "://" + host + ":" + std::to_string(port) + "/";
+  }
+
+  std::shared_ptr<opentelemetry::ext::http::client::Request> CreateRequest() noexcept override
+  {
+    http_request_.reset(new Request());
+    return http_request_;
+  }
+
+  MOCK_METHOD(void,
```

Comment:
Do we need to mock the SendRequest(), or can we create a empty (200 OK) response, and call the callback from same thread. Something like this 
```cpp
auto response   = std::unique_ptr<Response>(new Response());
.. populate response object..
callback.OnResponse(response);
```

This will allow us to remove dependency of gtest/gmock from this target?

Reply:
I have used Mock for testing the request that is going to be sent. The other option would be to make the client test itself via:
```cpp
std::function<void(http_client::nosend::Session *)> CreateJsonTester(
    const std::string report_trace_id)
{
  return [report_trace_id](http_client::nosend::Session *mock_session) {
    auto check_json    = nlohmann::json::parse(mock_session->GetRequest()->body_, nullptr, false);
    // extract received_trace_id from request
    EXPECT_EQ(received_trace_id, report_trace_id);
  };
}
```

Then store this function into session, so whenever there is a request going to be sent the session tests itself:
```cpp
  void SendRequest(opentelemetry::ext::http::client::EventHandler &callback) noexcept override
  {
    if (test_input_)
    {
      test_input_(this);
    }
    // let the otlp_http_client to continue
    Response response;
    callback.OnResponse(response);
  }
```
Shall I keep gMock or let the session test itself?

Reply:
Thank. Let's keep the gMock for now, we can fix it later.

I am just thinking another solution to change signature for [EventHandler::OnEvent](https://github.com/open-telemetry/opentelemetry-cpp/blob/ef3a623a3bedd3fded167bade625d7fa392d2b88/ext/include/opentelemetry/ext/http/client/http_client.h#L203) to something like:

`virtual void OnEvent(SessionState, void* data = nullptr, size_t size = 0)) noexcept = 0;`

The method in its current form is not usable. Sending back client implementation-specific `data` would be usable in real-world scenarios like server certificate validation during SSL Handshake. And NoSend client can then use this to get the Request object to be tested, something like

```
void SendRequest(EventHandler &callback) noexcept override
{
    auto response   = std::unique_ptr<Response>(new Response());
    callback.OnEvent(EventState::Connected, (void*)http_request_.get(), size);    
      .. populate response object..
    callback.OnResponse(response);
}
```

But don't want to delay this PR more, let's have this fix separately. We can have a github ticket to add this support / or probably remove mock from nosend client.